packages/dds/tree/src/test/simple-tree/simpleTree.bench.ts

```diff
@@ -80,6 +82,266 @@ describe("SimpleTree benchmarks", () => {
+						read: (tree: MySchema) => tree.leafUnion as number,
+						expected: 1,
+					},
+					{
+						title: `Read value from union of leaf and non-leaf`,
+						initUnhydrated: () => new MySchema({ complexUnion: 1 }),
+						initFlex: () => hydrate(MySchema, { complexUnion: 1 }),
+						read: (tree: MySchema) => tree.complexUnion as number,
+						expected: 1,
+					},
+					{
+						title: `Read undefined from leaf`,
+						initUnhydrated: () => new MySchema({}),
+						initFlex: () => hydrate(MySchema, {}),
+						read: (tree: MySchema) => tree.value,
+						expected: undefined,
+					},
+					{
+						title: `Read undefined from union of two leaves`,
+						initUnhydrated: () => new MySchema({}),
+						initFlex: () => hydrate(MySchema, {}),
+						read: (tree: MySchema) => tree.leafUnion as number,
+						expected: undefined,
+					},
+					{
+						title: `Read undefined from union of leaf and non-leaf`,
+						initUnhydrated: () => new MySchema({}),
+						initFlex: () => hydrate(MySchema, {}),
+						read: (tree: MySchema) => tree.complexUnion as number,
+						expected: undefined,
+					},
+				];
+
+				for (const { title, initUnhydrated, initFlex, read, expected } of testCases) {
+					testAccessToLeaf(`${title} (unhydrated node)`, initUnhydrated, read, expected);
+					testAccessToLeaf(`${title} (flex node)`, initFlex, read, expected);
+				}
+			});
+
+			describe("Required object property", () => {
+				const factory = new SchemaFactory("test");
+				class MySchema extends factory.object("root", {
+					value: factory.number,
+					leafUnion: [factory.number, factory.string],
+					complexUnion: [
+						factory.number,
+						factory.object("inner", {
+							value: factory.number,
+						}),
+					],
+				}) {}
+
+				const testCases = [
+					{
+						title: `Read value from leaf`,
+						initUnhydrated: () => new MySchema({ value: 1, leafUnion: 1, complexUnion: 1 }),
+						initFlex: () => hydrate(MySchema, { value: 1, leafUnion: 1, complexUnion: 1 }),
+						read: (tree: MySchema) => tree.value,
+						expected: 1,
+					},
+					{
+						title: `Read value from union of two leaves`,
+						initUnhydrated: () => new MySchema({ value: 1, leafUnion: 1, complexUnion: 1 }),
+						initFlex: () => hydrate(MySchema, { value: 1, leafUnion: 1, complexUnion: 1 }),
+						read: (tree: MySchema) => tree.leafUnion as number,
+						expected: 1,
+					},
+					{
+						title: `Read value from union of leaf and non-leaf`,
+						initUnhydrated: () => new MySchema({ value: 1, leafUnion: 1, complexUnion: 1 }),
+						initFlex: () => hydrate(MySchema, { value: 1, leafUnion: 1, complexUnion: 1 }),
+						read: (tree: MySchema) => tree.complexUnion as number,
+						expected: 1,
+					},
+				];
+
+				for (const { title, initUnhydrated, initFlex, read, expected } of testCases) {
+					testAccessToLeaf(`${title} (unhydrated node)`, initUnhydrated, read, expected);
+					testAccessToLeaf(`${title} (flex node)`, initFlex, read, expected);
+				}
```

Comment:
this logic seems duplicated.

Perhaps you can hoisted it further out and apply it to all of these tests at once rather than duplicating it for each batch of tests.